Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs

```diff
@@ -1,20 +1,239 @@
+            if (message.PillType > 19)
+                return;
+
+            chemMaster.PillType = message.PillType;
+            UpdateUiState(chemMaster);
+            ClickSound(chemMaster);
+        }
+
+        private void OnReagentButtonMessage(EntityUid uid, ChemMasterComponent chemMaster, ChemMasterReagentAmountButtonMessage message)
+        {
+            // Ensure the amount corresponds to one of the reagent amount buttons.
+            if (!Enum.IsDefined(typeof(ChemMasterReagentAmount), message.Amount))
+                return;
+
+            switch (chemMaster.Mode)
+            {
+                case ChemMasterMode.Transfer:
+                    TransferReagents(chemMaster, message.ReagentId, message.Amount.GetFixedPoint(), message.FromBuffer);
+                    break;
+                case ChemMasterMode.Discard:
+                    DiscardReagents(chemMaster, message.ReagentId, message.Amount.GetFixedPoint(), message.FromBuffer);
+                    break;
+                default:
+                    // Invalid mode.
+                    return;
+            }
+
+            ClickSound(chemMaster);
+        }
+
+        private void TransferReagents(ChemMasterComponent chemMaster, string reagentId, FixedPoint2 amount, bool fromBuffer)
+        {
+            var containerSlot = CompOrNull<ItemSlotsComponent>(chemMaster.Owner)?.Slots.GetValueOrDefault("container_slot");
+            if (containerSlot is null)
+                return;
+
+            if (!containerSlot.HasItem ||
+                containerSlot.Item is not {Valid: true} container ||
+                !TryComp(container, out FitsInDispenserComponent? fits) ||
+                !_solutionContainerSystem.TryGetSolution(container, fits.Solution, out var containerSolution)
+            )
+                return;
+
+            var bufferSolution = _solutionContainerSystem.EnsureSolution(chemMaster.Owner, ChemMasterComponent.SolutionName);
+
+            if (fromBuffer) // Buffer to container
+            {
+                amount = FixedPoint2.Min(amount, containerSolution.AvailableVolume);
+                amount = bufferSolution.RemoveReagent(reagentId, amount);
+                _solutionContainerSystem.TryAddReagent(container, containerSolution, reagentId, amount, out var _);
+            }
+            else // Container to buffer
+            {
+                amount = FixedPoint2.Min(amount, containerSolution.GetReagentQuantity(reagentId));
+                _solutionContainerSystem.TryRemoveReagent(container, containerSolution, reagentId, amount);
+                bufferSolution.AddReagent(reagentId, amount);
+            }
+
+            UpdateUiState(chemMaster, updateLabel: true);
+        }
+
+        private void DiscardReagents(ChemMasterComponent chemMaster, string reagentId, FixedPoint2 amount, bool fromBuffer)
+        {
+            var bufferSolution = _solutionContainerSystem.EnsureSolution(chemMaster.Owner, ChemMasterComponent.SolutionName);
```

Comment:
do you really need to ensure the solution every time? just get the solution and early out if it doesn't exist at this point